lib/crypto/tasks.py

```diff
@@ -1,18 +1,102 @@
+import json
 import logging
+import shutil
+import zipfile
 
 from celeryutils import task
+from lxml import etree
 
+import amo
 from versions.models import Version
 from lib.crypto.packaged import SigningError
 
+# Python 2.6 and earlier doesn't have context manager support
+ZipFile = zipfile.ZipFile
+if not hasattr(zipfile.ZipFile, "__enter__"):
+    class ZipFile(zipfile.ZipFile):
+        def __enter__(self):
+            return self
+
+        def __exit__(self, type, value, traceback):
+            self.close()
+
 log = logging.getLogger('z.task')
 
 
 @task
 def sign_addons(addon_ids, **kw):
-    log.info('[%s] Signing addons.' % len(addon_ids))
-    for version in Version.objects.filter(addon_id__in=addon_ids):
+    log.info('[{0}] Signing addons.'.format(addon_ids))
+    for version in Version.objects.filter(
+            addon_id__in=addon_ids, addon__type=amo.ADDON_EXTENSION):
+        log.info('Signing addon {0}, version {1}'.format(version.addon,
+                                                         version))
+        # We need to bump the version number of the file and the version, so
+        # the Firefox extension update mecanism picks this new signed version
+        # and installs it.
+        bump_version_number(version)
         try:
             version.sign_files()
         except SigningError as e:
-            log.warning('Failed signing version %s: %s.' % (version.pk, e))
+            log.warning(
+                'Failed signing version {0}: {1}.'.format(version.pk, e))
+
+
+def bump_version_number(version):
+    """Add a .1 to the version number in the install.rdf or package.json."""
+    for file_obj in version.files.all():  # Bump the numbers in each file.
+        # Create a new xpi with the bumped version.
+        bumped = '{0}.bumped'.format(file_obj.file_path)
+        # Copy the original XPI, with the updated install.rdf or package.json.
+        with ZipFile(file_obj.file_path, 'r') as source:
+            file_list = source.infolist()
+            with ZipFile(bumped, 'w', zipfile.ZIP_DEFLATED) as dest:
+                for file_ in file_list:
+                    content = source.read(file_.filename)
+                    if file_.filename == 'install.rdf':
+                        content = _bump_version_in_install_rdf(content)
+                    if file_.filename == 'package.json':
+                        content = _bump_version_in_package_json(content)
+                    dest.writestr(file_, content)
+        # Move the bumped file to the original file.
+        shutil.move(bumped, file_obj.file_path)
+    # Now update the Version model.
+    version.update(version='{0}.1'.format(version.version))
+
+
+def _dot_one(version):
+    """Returns the version with an appended .1 on it."""
+    return '{0}.1'.format(version)
```

Comment:
This isn't going to work. See https://bugzil.la/1070227#c3

For version numbers which already have 4 components, this is going to make the new version number *lower* than the current version number. This turned out to be a significant issue in last round of SDK repacks.